lwtime/nattime.go

```diff
@@ -0,0 +1,198 @@
+package lwtime
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type naturalAdjective string
+
+const (
+	// both plural and singular regular expressions should contain 3 capture groups
+	naturalPluralRE   = `^(last)\s(\d+)\s(years|months|weeks|days|hours|minutes|seconds)$`
+	naturalSingularRE = `^(this|current|previous|last)(\s)(year|month|week|day|hour|minute|second)$`
+
+	Today     naturalAdjective = "today"
+	Yesterday naturalAdjective = "yesterday"
+	This      naturalAdjective = "this"
+	Current   naturalAdjective = "current"
+	Previous  naturalAdjective = "previous"
+	Last      naturalAdjective = "last"
+)
+
+func (nta naturalAdjective) isValid() bool {
+	switch naturalAdjective(strings.ToLower(string(nta))) { // inline lowercase conversion
+	case Today, Yesterday, This, Current, Previous, Last:
+		return true
+	}
+	return false
+}
+
+type natural struct {
+	adjective naturalAdjective
+	num       string
+	iNum      int
+	unit      relativeUnit
+}
+
+func (nt *natural) loadRelativeUnit(u string) bool {
+	switch strings.ToLower(u) {
+	case "year", "years":
+		nt.unit = Year
+	case "month", "months":
+		nt.unit = Month
+	case "week", "weeks":
+		nt.unit = Week
+	case "day", "days":
+		nt.unit = Day
+	case "hour", "hours":
+		nt.unit = Hour
+	case "minute", "minutes":
+		nt.unit = Minute
+	case "second", "seconds":
+		nt.unit = Second
+	default:
+		return false
+	}
+	return true
+}
+
+func newNatural(s string) (natural, error) {
+	nt := natural{}
+	s = strings.ToLower(s)
+
+	// Today
+	if naturalAdjective(s) == Today {
+		nt.adjective = This
+		nt.num = "1"
+		nt.iNum = 1
+		nt.unit = relativeUnit("d")
+		return nt, nil
+	}
+	// Yesterday
+	if naturalAdjective(s) == Yesterday {
+		nt.adjective = Previous
+		nt.num = "1"
+		nt.iNum = 1
+		nt.unit = relativeUnit("d")
+		return nt, nil
+	}
+	// Singular
+	var nt_parts []string
+	singularRE := regexp.MustCompile(naturalSingularRE)
+	if nt_parts = singularRE.FindStringSubmatch(s); s == "" || nt_parts == nil {
+		// Plural
+		pluralRE := regexp.MustCompile(naturalPluralRE)
+		if nt_parts = pluralRE.FindStringSubmatch(s); s == "" || nt_parts == nil {
+			return nt, errors.New(fmt.Sprintf("natural time (%s) is invalid", s))
+		}
+	}
+	// Adjective
+	nt.adjective = naturalAdjective(nt_parts[1])
+	if !nt.adjective.isValid() {
+		// this would indicate a code mismatch between enumerated adjectives and regex
+		return nt, errors.New(fmt.Sprintf("invalid adjective for natural time (%s)", s))
+	}
+	// Num
+	nt.num = nt_parts[2]
+	var err error
+	nt.iNum, err = strconv.Atoi(nt.num)
+	if err != nil {
+		nt.num = "1"
+		nt.iNum = 1
+	}
+	// Unit
+	if ok := nt.loadRelativeUnit(nt_parts[3]); !ok {
+		// this would indicate a code mismatch between relative units and regex
+		return nt, errors.New(fmt.Sprintf("invalid unit for natural time (%s)", s))
+	}
+	return nt, nil
+}
+
+func (nt natural) range_(t time.Time) (start time.Time, end time.Time, err error) {
```

Comment:
can we call this function `range` only? the underscore `_` is used as a black identifier to denote that "things are not being used"
```suggestion
func (nt natural) range(t time.Time) (start time.Time, end time.Time, err error) {
```

Reply:
I think `range` is a built-in so Go was complaining about this.  I can just find a different name for it since it's internal anyway...

Reply:
Fixed.